<center>
<img src="../../img/ods_stickers.jpg">
## Открытый курс по машинному обучению. Сессия № 2
</center>
Автор материала: программист-исследователь Mail.ru Group, старший преподаватель Факультета Компьютерных Наук ВШЭ Юрий Кашницкий. Материал распространяется на условиях лицензии [Creative Commons CC BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/). Можно использовать в любых целях (редактировать, поправлять и брать за основу), кроме коммерческих, но с обязательным упоминанием автора материала.

# <center> Домашнее задание № 1
## <center> Анализ данных по сердечно-сосудистым заболеваниям
*Идеальный просмотр Jupyter notebooks возможен только локально, GitHub и nbviewer могут неправильно отбражать формулы или картинки.*

**В задании предлагается с помощью Pandas ответить на несколько вопросов по данным о сердечно-сосудистых заболеваниях (качать данные не надо, они уже есть в репозитории). Данные использовались в соревновании [Ml Boot Camp 5](http://mlbootcamp.ru/round/12/sandbox/).**

**Заполните код в клетках (где написано "Ваш код здесь") и ответьте на вопросы в [веб-форме](https://goo.gl/8D8uDf).**

В соревновании предлагалось определить наличие/отсутствие сердечно-сосудистых заболеваний (ССЗ) по результатам осмотра пациента.

**Описание данных.**

Объективные признаки:

 - Возраст (age)
 - Рост (height)
 - Вес (weight)
 - Пол (gender)
 

Результаты измерения:

 - Артериальное давление верхнее и нижнее (ap_hi, ap_lo)
 - Холестерин (cholesterol)
 - Глюкоза (gluc)
 

Субъективные признаки (со слов пациентов):

 - Курение (smoke)
 - Употребление алкоголя (alco)
 - Физическая активность (active)
 
Целевой признак (который интересно будет прогнозировать):
 - Наличие сердечно-сосудистых заболеваний по результатам классического врачебного осмотра (cardio)

Значения показателей холестерина и глюкозы представлены одним из трех классов: норма, выше нормы, значительно выше нормы. Значения субъективных признаков — бинарны.

Все показатели даны на момент осмотра.

Мы будем работать только с обучающей выборкой и с помощью `Pandas` проведем первичный анализ данных. 

Из библиотек нам понадобился только Pandas.

In [45]:
import pandas as pd

Считываем данные из CSV-файла в объект pandas DataFrame.  

In [46]:
df = pd.read_csv('../../data/mlbootcamp5_train.csv', 
                 sep=';', index_col='id')

Посмотрим на первые 5 записей.

In [47]:
df.head()

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168      62    110     80            1     1      0   
1   20228       1     156      85    140     90            3     1      0   
2   18857       1     165      64    130     70            3     1      0   
3   17623       2     169      82    150    100            1     1      0   
4   17474       1     156      56    100     60            1     1      0   

    alco  active  cardio  
id                        
0      0       1       0  
1      0       1       1  
2      0       0       1  
3      0       1       1  
4      0       0       0

**<font color='red'>Вопрос 1 (1 балл).</font> Сколько мужчин и женщин представлено в этом наборе данных? Не было дано расшифровки признака "пол" (какому полу соответствует 1, а какому – 2 в признаке ``gender``) – это определите посмотрев также на рост при разумном предположении, что в среднем мужчины выше.**

**<font color='red'>Варианты:</font>**
- **<font color='green'>45530 женщин и 24470 мужчин</font>**
- 45530 мужчин и 24470 женщин
- 45470 женщин и 24530 мужчин
- 45470 мужчин и 24530 женщин

Посчитаем средний рост по различным значениям пола:

In [48]:
df.groupby('gender')['height'].mean()

gender
1    161.355612
2    169.947895
Name: height, dtype: float64

In [75]:
df.pivot_table(['height','smoke'],['gender'],aggfunc='mean')

height     smoke
gender                      
1       161.355612  0.017856
2       169.947895  0.218880

In [49]:
df['gender_norm'] = df['gender'].map(lambda g: 'male' if g == 2 else 'female')
df.head()

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168      62    110     80            1     1      0   
1   20228       1     156      85    140     90            3     1      0   
2   18857       1     165      64    130     70            3     1      0   
3   17623       2     169      82    150    100            1     1      0   
4   17474       1     156      56    100     60            1     1      0   

    alco  active  cardio gender_norm  
id                                    
0      0       1       0        male  
1      0       1       1      female  
2      0       0       1      female  
3      0       1       1        male  
4      0       0       0      female

161 и почти 170 см в среднем, делаем вывод, что gender=1 обозначает женщин, а gender=2 – мужчин. 

In [50]:
df.gender_norm.value_counts()

female    45530
male      24470
Name: gender_norm, dtype: int64

То есть в выборке 45530 женщин и 24470 мужчин.

**<font color='red'>Вопрос 2 (1 балл).</font> Кто в среднем чаще указывает, что употребляет алкоголь – мужчины или женщины?**

**<font color='red'>Варианты:</font>**
- **<font color='green'>мужчины</font>** 
- женщины

In [51]:
df.groupby('gender_norm')['alco'].mean()

gender_norm
female    0.025500
male      0.106375
Name: alco, dtype: float64

Конечно, мужчины.

**<font color='red'>Вопрос 3 (1 балл).</font> Во сколько раз (округленно, `round`) процент курящих среди мужчин больше, чем процент курящих среди женщин (по крайней мере, по этим анкетным данным)?**

**<font color='red'>Варианты:</font>**
- 4
- 8
- **<font color='green'>12</font>**
- 16

In [52]:
df.groupby('gender_norm')['smoke'].mean()

gender_norm
female    0.017856
male      0.218880
Name: smoke, dtype: float64

In [53]:
df[df['gender_norm'] == 'male']['smoke'].mean() / df[df['gender_norm'] == 'female']['smoke'].mean()

12.257833097333833

Ого! В 12 раз, почти 22% у мужчин и только 1.7% у женщин.

**<font color='red'>Вопрос 4 (1 балл).</font> Вы наверняка заметили, что значения возраста какие-то странные. Догадайтесь, в чем здесь измеряется возраст, и ответьте, на сколько месяцев (примерно) отличаются медианные значения возраста курящих и некурящих.**

**<font color='red'>Варианты:</font>**
- 5
- 10
- 15
- **<font color='green'>20</font>**

Возраст здесь указан в днях. 

In [54]:
df['age_norm'] = df['age'] / 365.25
df.head()

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168      62    110     80            1     1      0   
1   20228       1     156      85    140     90            3     1      0   
2   18857       1     165      64    130     70            3     1      0   
3   17623       2     169      82    150    100            1     1      0   
4   17474       1     156      56    100     60            1     1      0   

    alco  active  cardio gender_norm   age_norm  
id                                               
0      0       1       0        male  50.357290  
1      0       1       1      female  55.381246  
2      0       0       1      female  51.627652  
3      0       1       1        male  48.249144  
4      0       0       0      female  47.841205

In [55]:
df.groupby('smoke')['age_norm'].median()

smoke
0    53.995893
1    52.361396
Name: age_norm, dtype: float64

Медианный возраст курящих – 52.4 года, некурящих – 54. Можно точно не считать, видно, что ответ на вопрос – 20 месяцев. Если все же точно, то это можно так посчитать:

In [56]:
(df[df['smoke'] == 0]['age_norm'].median() - 
 df[df['smoke'] == 1]['age_norm'].median()) * 12

19.61396303901438

**<font color='red'>Вопрос 5 (2 балла).</font> В [статье](https://ru.wikipedia.org/wiki/Сердечно-сосудистый_риск) на Википедии про сердечно-сосудистый риск показана шкала SCORE для расчёта риска смерти от сердечно-сосудистого заболевания в ближайшие 10 лет. Вот она:**
<img src='../../img/SCORE2007.png' width=70%>

Давайте посмотрим на правый верхний прямоугольник, отображающий сегмент курящих мужчин в возрасте от 60 до 64 лет включительно. (Неочевидно, но тут для возраста и давления цифры означают верхнюю границу, и она не включается).

Видим 9-ку в левом нижнем углу этого прямоугольника и 47 – в правом верхнем. То есть если при этом систолическое (т.е. верхнее) артериальное давление – меньше 120 мм рт.ст., а уровень холестерина – 4 ммоль/л, то риск ССЗ оценивается примерно в 5 раз ниже, чем если бы давление лежало в интервале [160, 180), а холестерина было бы 8 ммоль/л.

Давайте посчитаем аналогичное значение, но на наших данных. 

Уточнения:
- Посчитайте признак ``age_years`` – возраст в годах, округлив до целых (`round`). Для данного примера отберите курящих мужчин от 60 до 64 лет включительно
- Категории уровня холестрина на рисунке и в наших данных отличаются. Отображение значений на картинке в значения признака ``cholesterol`` следующее: 4 ммоль/л $\rightarrow$ 1, 5-7 ммоль/л $\rightarrow$ 2, 8 ммоль/л $\rightarrow$ 3.
- Интересуют 2 подвыборки курящих мужчин возраста от 60 до 64 лет: первая с верхним артериальным давлением меньше 120 мм рт.ст. и концентрацией холестерина – 4 ммоль/л, а вторая – с верхним артериальным давлением от 160 до 180 мм рт.ст. (не включительно) и концентрацией холестерина – 8 ммоль/л.

Во сколько раз (округленно, `round`) отличаются доли больных людей в этих двух подвыборках? Посчитайте на наших данных.

**<font color='red'>Варианты:</font>**
- 2
- **<font color='green'>3</font>**
- 4
- 5

In [57]:
df['age_years'] = (df['age'] / 365.25).round().astype('int')
df.head()

age  gender  height  weight  ap_hi  ap_lo  cholesterol  gluc  smoke  \
id                                                                          
0   18393       2     168      62    110     80            1     1      0   
1   20228       1     156      85    140     90            3     1      0   
2   18857       1     165      64    130     70            3     1      0   
3   17623       2     169      82    150    100            1     1      0   
4   17474       1     156      56    100     60            1     1      0   

    alco  active  cardio gender_norm   age_norm  age_years  
id                                                          
0      0       1       0        male  50.357290         50  
1      0       1       1      female  55.381246         55  
2      0       0       1      female  51.627652         52  
3      0       1       1        male  48.249144         48  
4      0       0       0      female  47.841205         48

In [58]:
df['age_years'].max()

65

Самые старые в выборке – как раз 65-летние. Совпадение? Не думаю! Отберем курящих мужчин от 60 до 64 лет.

In [59]:
smoking_old_men = df[(df['gender_norm'] == 'male') & (df['age_years'] >= 60)
                    & (df['age_years'] < 65) & (df['smoke'] == 1)]

Если в этом сегменте уровень холестерина 1, а верхнее артериальное давление – до 120 мм рт.ст., то больных – 26%. 

In [60]:
smoking_old_men[(smoking_old_men['cholesterol'] == 1) &
               (smoking_old_men['ap_hi'] < 120)]['cardio'].mean()

0.2631578947368421

Если же в этом сегменте уровень холестерина 3, а верхнее артериальное давление – от 160 до 180 мм рт.ст., то больных – 89%. 

In [61]:
smoking_old_men[(smoking_old_men['cholesterol'] == 3) &
               (smoking_old_men['ap_hi'] >= 160) &
               (smoking_old_men['ap_hi'] < 180)]['cardio'].mean()

0.8636363636363636

Итого разница примерно в 3 раза. Не в 5, как говорит шкала SCORE, но возможно, оцениваемый ей риск ССЗ нелинейно зависит от процента больных в данном сегменте. 

**<font color='red'>Вопрос 6 (2 балла).</font> Постройте новый признак – BMI ([Body Mass Index](https://en.wikipedia.org/wiki/Body_mass_index)). Для этого надо вес в килограммах поделить на квадрат роста в метрах. Нормальными считаются значения BMI от 18.5 до 25. Выберите верные утверждения.**

**<font color='red'>Утверждения:</font>**
- Медианный BMI по выборке лежит в пределах нормы
- **<font color='green'>У женщин в среднем BMI выше, чем у мужчин</font>**
- У здоровых в среднем BMI выше, чем у больных
- **<font color='green'>В сегменте здоровых и непьющих мужчин BMI ближе к норме, чем в сегменте здоровых и непьющих женщин</font>**

In [62]:
df['BMI'] = df['weight'] / (df['height'] / 100) ** 2
df['BMI'].median()

26.374068120774975

Первое утверждение неверно, медианный BMI в этой выборке превышает норму в 25 единиц.

In [63]:
df.groupby('gender_norm')['BMI'].median()

gender_norm
female    26.709402
male      25.910684
Name: BMI, dtype: float64

Второе утверждение верно, у женщин в среднем BMI выше.

In [64]:
df.groupby('cardio')['BMI'].median()

cardio
0    25.476660
1    27.471384
Name: BMI, dtype: float64

Третье утверждение неверно.

In [65]:
df.groupby(['gender', 'alco', 'cardio'])['BMI'].median()

gender  alco  cardio
1       0     0         25.654372
              1         27.885187
        1     0         27.885187
              1         30.110991
2       0     0         25.102391
              1         26.674874
        1     0         25.351541
              1         27.530797
Name: BMI, dtype: float64

Последнее утверждение верно.

**<font color='red'>Вопрос 7 (2 балла).</font> ** Можно заметить, что данные не особо-то чистые, много в них всякой "грязи" и неточностей. Еще лучше мы это увидим, когда обсудим визуализацию данных.

Отфильтруйте следующие сегменты пациентов (считаем это ошибками в данных)
- указанное нижнее значение артериального давления выше верхнего
- рост строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили (используйте `pd.Series.quantile`, если не знаете, что это такое – прочитайте)
- вес строго меньше 2.5%-перцентили или строго больше 97.5%-перцентили

Этот вовсе не вся чистка данных, которую можно было проделать, но пока остановимся на этом.

Сколько процентов данных (округленно, `round`) мы выбросили?

**<font color='red'>Варианты:</font>**
- 8
- 9
- **<font color='green'>10</font>**
- 11

In [66]:
filtered_df = df[(df['ap_lo'] <= df['ap_hi']) & 
                 (df['height'] >= df['height'].quantile(0.025)) &
                 (df['height'] <= df['height'].quantile(0.975)) &
                 (df['weight'] >= df['weight'].quantile(0.025)) & 
                 (df['weight'] <= df['weight'].quantile(0.975))]
print(filtered_df.shape[0] / df.shape[0])

0.9037


Выбросили около 10% исходных данных.

### Дополнительные эксперименты с данными

Сводная таблица (пол, больные)

In [74]:
pd.crosstab(df['cardio'], df['gender_norm'])

gender_norm  female   male
cardio                    
0             22914  12107
1             22616  12363